src/Plutarch/Test/Export.hs

```diff
@@ -0,0 +1,728 @@
+ policies, the redeemer is ignored; you can set it to '()' to keep the
+ compiler happy.
+
+ @since 1.2.1
+-}
+data ScriptParams (dat :: Type) (red :: Type) = ScriptParams
+  { datum :: dat
+  , redeemer :: red
+  , context :: ScriptContext
+  }
+
+makeFieldLabelsNoPrefix ''ScriptParams
+
+{- | A combination of (initial) script parameters, as well as any required
+ parameters that should be passed to the linker when assembling tests.
+
+ @since 1.2.1
+-}
+data TestParams (dat :: Type) (red :: Type) (p :: Type) = TestParams
+  { scriptParams :: ScriptParams dat red
+  , linkerParams :: p
+  }
+
+makeFieldLabelsNoPrefix ''TestParams
+
+{- | What you expect to happen when a script is tested with specified
+ parameters.
+
+ @since 1.2.1
+-}
+data ScriptOutcome = Crashes | Runs
+  deriving stock
+    ( -- | @since 1.2.1
+      Eq
+    , -- | @since 1.2.1
+      Ord
+    , -- | @since 1.2.1
+      Show
+    )
+
+{- | A helper newtype for passing script arguments implicitly, as well as
+ modifying them using local scoping.
+
+ The intended use is with @do@-notation, using a combination of contextual
+ argument modifiers and assertions, then \'running\' with 'exportTests':
+
+ > tests :: TestTree
+ > tests = exportTests "my tests" export linker testParams $ do
+ >    assertValidator Crashes "my context is bad" "script1"
+ >    modifyingSC fixUpContext $ do
+ >      assertValidator Crashes "context fine, redeemer bad" "script1"
+ >      modifyingRedeemer fixUpRedeemer $ do
+ >        assertValidator Runs "now everything is fine" "script1"
+ >      -- Redeemer fix-up resets here
+ >      assertMintingPolicy Crashes "inappropriate datum" "script2"
+ >      settingDatum' redeemerDatum $ do
+ >        assertMintingPolicy Runs "appropriate datum" "script2"
+
+ You can also define separate \'bits\' of a suite using 'WithExport' directly.
+
+ @since 1.2.1
+-}
+newtype WithExport (dat :: Type) (red :: Type) (info :: Type) (a :: Type)
+  = WithExport (RWS (WithExportEnv dat red info) WithExportLog () a)
+  deriving
+    ( -- | @since 1.2.1
+      Functor
+    , -- | @since 1.2.1
+      Applicative
+    , -- | @since 1.2.1
+      Monad
+    )
+    via ( RWS (WithExportEnv dat red info) WithExportLog ()
+        )
+
+{- | Given a name for the overall test suite, a raw export, a linker, and some
+ test parameters, assemble a 'TestTree'.
+
+ This acts as a 'testGroup' replacement, and can be used in a similar way. The
+ exact number of tests this will generate depends on the script export given,
+ and the scripts we ask to test:
+
+ - If linking fails, you will get a single, always-failing, test, which
+ presents the linker error;
+ - If any tests request non-existent scripts (that is, a name that doesn't
+ correspond to an existing script), those requests will be bundled into a
+ single, always-failing, test, which presents the name that doesn't have a
+ script associated with it;
+ - Otherwise, one test per request.
+
+ @since 1.2.1
+-}
+exportTests ::
```

Comment:
We need handler for `ScriptExport` as well. While raw script one should be the default, we still need to test `liqwid-onchain` which exports directly to `ScriptExport`. 

I suggest making `WithExportEnv` have `Maybe RawScriptExport`, so that way we can automatically omit the script role check.

Reply:
So we would have _two_ eliminators for `WithExport` I guess: one for raw, one for compiled?